src/solver/impls/snes/snes.cxx

```diff
@@ -86,33 +127,457 @@ int SNESSolver::init(int nout, BoutReal tstep) {
+
+     */
+    MatCreateSNESMF(snes, &Jmf);
+
+    // Set a function to be called for differencing
+    // This can be a linearised form of the SNES function
+    MatMFFDSetFunction(Jmf, FormFunctionForDifferencing, this);
+
+    // Calculate Jacobian matrix free using FormFunctionForDifferencing
+    SNESSetJacobian(snes, Jmf, Jmf, MatMFFDComputeJacobian, this);
+
+  } else {
+    // Calculate the Jacobian using finite differences
+
+    bool use_coloring = (*options)["use_coloring"]
+                            .doc("Use matrix coloring to calculate Jacobian?")
+                            .withDefault<bool>(true);
+
+    if (use_coloring) {
+      // Use matrix coloring
+      // This greatly reduces the number of times the rhs() function needs
+      // to be evaluated when calculating the Jacobian.
+
+      // Use global mesh for now
+      Mesh* mesh = bout::globals::mesh;
+
+      //////////////////////////////////////////////////
+      // Get the local indices by starting at 0
+      Field3D index = globalIndex(0);
+
+      //////////////////////////////////////////////////
+      // Pre-allocate PETSc storage
+
+      int localN = getLocalN(); // Number of rows on this processor
+      int n2d = f2d.size();
+      int n3d = f3d.size();
+
+      // Set size of Matrix on each processor to localN x localN
+      MatCreate(BoutComm::get(), &Jmf);
+      MatSetSizes(Jmf, localN, localN, PETSC_DETERMINE, PETSC_DETERMINE);
+      MatSetFromOptions(Jmf);
+
+      PetscInt *d_nnz, *o_nnz;
+      PetscMalloc((localN) * sizeof(PetscInt), &d_nnz);
+      PetscMalloc((localN) * sizeof(PetscInt), &o_nnz);
+
+      // Set values for most points
+      if (mesh->LocalNz > 1) {
+        // A 3D mesh, so need points in Z
+
+        for (int i = 0; i < localN; i++) {
+          // Non-zero elements on this processor
+          d_nnz[i] = 7 * n3d + 5 * n2d; // Star pattern in 3D
+          // Non-zero elements on neighboring processor
+          o_nnz[i] = 0;
+        }
+      } else {
+        // Only one point in Z
+
+        for (int i = 0; i < localN; i++) {
+          // Non-zero elements on this processor
+          d_nnz[i] = 5 * (n3d + n2d); // Star pattern in 2D
+          // Non-zero elements on neighboring processor
+          o_nnz[i] = 0;
+        }
+      }
+
+      // X boundaries
+      if (mesh->firstX()) {
+        // Lower X boundary
+        for (int y = mesh->ystart; y <= mesh->yend; y++) {
+          for (int z = 0; z < mesh->LocalNz; z++) {
+            int localIndex = ROUND(index(mesh->xstart, y, z));
+            ASSERT2((localIndex >= 0) && (localIndex < localN));
+            if (z == 0) {
+              // All 2D and 3D fields
+              for (int i = 0; i < n2d + n3d; i++) {
+                d_nnz[localIndex + i] -= (n3d + n2d);
+              }
+            } else {
+              // Only 3D fields
+              for (int i = 0; i < n3d; i++) {
+                d_nnz[localIndex + i] -= (n3d + n2d);
+              }
```

Comment:
Assuming the loop body is supposed to be identical between the branches, here's a more concise way of writing this that avoids the repetition in the loop body: 

```suggestion
            const auto num_fields = (z == 0) ? n2d + n3d : n3d;
            for (int i = 0; i < num_fields; i++) {
              d_nnz[localIndex + i] -= (n3d + n2d);
            }
```

Reply:
I also wonder if these loop bodies could be wrapped up into functions and reused? Might cut down on this section a fair bit